### PR TITLE
[SPARK-11835] Adds a sidebar menu to MLlib's documentation

### PR DESCRIPTION
**Pull request by thunterdb**
This PR adds a sidebar menu when browsing the user guide of MLlib. It uses a YAML file to describe the structure of the documentation. It should be trivial to adapt this to the other projects.

screen shot 2015-11-18 at 4 46 12 pm

### CODE REVIEW

File: docs/_data/menu-ml.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+- text: Feature extraction, transformation, and selection
+  url: ml-features.html
+- text: Decision trees for classification and regression
+  url: ml-decision-tree.html
+- text: Ensembles
+  url: ml-ensembles.html
+- text: Linear methods with elastic net regularization
```
Comment (mengxr) [inline docs/_data/menu-ml.yaml]: `elastic-net`

File: docs/_data/menu-mllib.yaml [hunk 2]
```diff
@@ -0,0 +1,75 @@
+- text: Data types
+  url: /mllib-data-types.html
```
Comment (mengxr) [inline docs/_data/menu-mllib.yaml]: remove `/`

File: docs/_data/menu-mllib.yaml [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+- text: Basic statistics
+  url: /mllib-statistics.html
```
Comment (mengxr) [inline docs/_data/menu-mllib.yaml]: ditto

File: docs/_includes/nav-left-wrapper-ml.html [hunk 4]
```diff
@@ -0,0 +1,6 @@
+<div class="left-menu">
+    <h3>ML</h3>
```
Comment (mengxr) [inline docs/_includes/nav-left-wrapper-ml.html]: Shall we use `spark.ml package` and `spark.mllib package` instead?

File: docs/_includes/nav-left-wrapper-ml.html [hunk 5]
```diff
@@ -0,0 +1,6 @@
...
+    {% include nav-left.html nav=include.nav-ml %}
+    <h3>MLlib</h3>
+    {% include nav-left.html nav=include.nav-mllib %}
+</div>
```
Comment (mengxr) [inline docs/_includes/nav-left-wrapper-ml.html]: nit: new line at the end

**Comment by mengxr** [comment]
Great work! Shall we hide the side-nav-bar if width is not sufficient? See attached screenshot:

screen shot 2015-11-18 at 6 45 11 pm

File: docs/_includes/nav-left-wrapper-ml.html (hunk 4 above)
Comment (thunterdb) [inline docs/_includes/nav-left-wrapper-ml.html]: sure

**Comment by thunterdb** [comment]
@mengxr with the fixes, the top menu breaks before the side menu :-)
screen shot 2015-11-19 at 11 30 13 am
screen shot 2015-11-19 at 11 30 25 am

**Comment by thunterdb** [comment]
This is about as much as I can do with my very limited knowledge of CSS, so additional fixes can be done in a separate PR.

**Comment by andrewor14** [comment]
nice! I wonder if there's something more general we can do. This is useful across all pages in the docs so it would be worth investigating if there are existing frameworks we can use. E.g. this format is nice: http://docs.scala-lang.org/overviews/core/futures.html

**Comment by thunterdb** [comment]
@andrewor14 this is a different issue: the SIPs show the table of contents within one document, which we already have with the `{:toc}` directive. This PR adds the per-project organization (linking various markdown files together). This is not something we can infer automatically because we show different levels of nesting for each of the pages. Each project already has this description in their overview pages, and I am proposing we move them to the side bar:
 - http://spark.apache.org/docs/latest/streaming-programming-guide.html
 - http://spark.apache.org/docs/latest/sql-programming-guide.html
 - http://spark.apache.org/docs/latest/graphx-programming-guide.html
 - http://spark.apache.org/docs/latest/sparkr.html

**Comment by thunterdb** [comment]
@mengxr based on our discussions, here is a slightly updated version:
 - menu is fixed
 - menu disappears under the page (but it looks half covered in the middle). Fixing that would require JS, I think (and sk-learn does not have a fixed menu):

screen shot 2015-11-19 at 3 10 50 pm

**Comment by mengxr** [comment]
@thunterdb Could you try the following CSS:

~~~css
.left-menu-wrapper {
  position: absolute;
  height: 100%;
  width: 256px;
  margin-top: -20px;
  padding-top: 20px;
  background-color: #F0F8FC;
}

.left-menu {
  position: fixed;
  width: 256px;
  padding-right: 10px;
}
~~~

It doesn't fix the issue with hiding the sidebar but makes the background lighter:

screen shot 2015-11-19 at 11 38 05 pm

**Comment by thunterdb** [comment]
@mengxr comment addressed

**Comment by mengxr** [comment]
test this please

**Comment by mengxr** [comment]
test this please

**Comment by mengxr** [comment]
test this please

**Comment by mengxr** [comment]
LGTM. Merged into master and branch-1.6. Thanks! (The Jenkins test failure is irrelevant. The PR build doesn't compile doc anyway. So I verified the changes on my local computer.)

**Pull request closed without merging**